src/main/scala/mesosphere/marathon/core/task/Task.scala

```diff
@@ -173,18 +173,139 @@ case class Task(taskId: Task.Id, runSpecVersion: Timestamp, status: Task.Status)
 object Task {
   private val log = LoggerFactory.getLogger(getClass)
 
-  case class Id(idString: String) extends Ordered[Id] {
+  /**
+    * Base for all task identifiers.
+    */
+  trait Id extends Ordered[Id] {
+    // A stringifed version of the id.
+    val idString: String
+
+    // Quick access to the underlying run spec identifier of the task.
+    val runSpecId: PathId
+
+    // Quick access to the underlying instance identifier of the task.
+    val instanceId: Instance.Id
+
+    val reservationId: String
+
+    // The Mesos task id representation of the task.
     lazy val mesosTaskId: MesosProtos.TaskID = MesosProtos.TaskID.newBuilder().setValue(idString).build()
-    lazy val runSpecId: PathId = Id.runSpecId(idString)
-    lazy val instanceId: Instance.Id = Id.instanceId(idString)
-    lazy val containerName: Option[String] = Id.containerName(idString)
+
+    // pre-instance-era executorId="marathon-$taskId" and compatibility reasons we need this calculation.
+    // Should be removed as soon as no tasks without instance exists (tbd)
+    def calculateLegacyExecutorId(): String = s"marathon-$idString"
+
+    override def compare(that: Id): Int = idString.compare(that.idString)
+
     /**
-      * For resident tasks, this will denote the number of launched tasks on a reservation since 1.5
+      * @return String representation for debugging.
       */
-    lazy val attempt: Option[Long] = Id.attempt(idString)
-
     override def toString: String = s"task [$idString]"
-    override def compare(that: Id): Int = idString.compare(that.idString)
+  }
+
+  /**
+    * A task identifier for legacy tasks. This cannot represent a pod instance. These ids won't be created but old Mesos
+    * tasks might have such an instance thus we have to represent them.
+    *
+    * This id can also represent an old reservation for a persistent app, ie an app with resident tasks.
+    *
+    * The ids match [[Task.Id.LegacyTaskIdRegex]].
+    *
+    * Examples:
+    *  - "myGroup_myApp.b6ff5fa5-7714-11e7-a55c-5ecf1c4671f6"
+    *  - "myGroup_myApp.b6ff5fa5-7714-11e7-a55c-5ecf1c4671f6"
+    *
+    * @param runSpecId Identifies the run spec the task was started with.
+    * @param separator This can be "." or "_".
+    * @param uuid A unique identifier of the task.
+    */
+  case class LegacyId(val runSpecId: PathId, separator: String, uuid: UUID) extends Id {
+
+    // A stringifed version of the id.
+    override val idString: String = runSpecId.safePath + separator + uuid
+
+    override lazy val instanceId: Instance.Id = Instance.Id(runSpecId, Instance.PrefixMarathon, uuid)
```

Comment:
This instance ID will never have a corresponding instance.

Reply:
Why not? Should it be `Instance.PrefixInstance` then?